### ISSUE TITLE
Put index where OpenDistro stores tenant data into attribute field of elasticsearch_opendistro_kibana_tenant 

### ISSUE THREAD

**Issue by nouseforaname** [phillbaker/terraform-provider-elasticsearch#169]
Hey,

if I want to create an index pattern in a specific tenant via elasticsearch_kibana_object
I need to know the index that was created for `elasticsearch_opendistro_kibana_tenant`

E.g. given a tenant name of `sec-test`
the created index is `.kibana_882685070_sectest_1`

Then I can do:

```
resource "elasticsearch_kibana_object" "test-pattern" {
   index = ".kibana_882685070_sectest_1"
   body = <<EOF
[
{
  "_id": "index-pattern:my-id",
  "_type": "doc",
  "_source": {
    "type": "index-pattern",
    "index-pattern": {
      "title": "my-pattern-*",
      "timeFieldName": "@timestamp"
    }
  }
}
]
EOF
```

The number and full index_name it calculates is reproducible with this snippet:

```
package main

import (
	"fmt"
	"regexp"
)

func main() {
	hashSum := int32(0)
	tenant := "sec-test"
	for _, char := range tenant {
		shift := (hashSum<<5)		
		hashSum = ( shift - hashSum) +int32(char-0);	
	}
	fmt.Println(hashSum)
	reg, err := regexp.Compile("[^a-zA-Z0-9]+")
    	if err != nil {
          fmt.Println("regexErr")
        }
	processedString := reg.ReplaceAllString(tenant, "")
	fmt.Printf(".kibana_%v_%v_1", hashSum,processedString )
}
```

I can create a PR for adding the index as an attribute to the  elasticsearch_opendistro_kibana_tenant resource if you want

**Comment by phillbaker** [comment]
> I can create a PR for adding the index as an attribute to the `elasticsearch_opendistro_kibana_tenant` resource if you want

Thanks - you're thinking of adding it as a computed attribute to `elasticsearch_opendistro_kibana_tenant`? What does the number at the end signify? Does it ever increase from 1?

**Comment by nouseforaname** [comment]
TBH, I currently do not know.. For now, before doing more research on this, I wanted to ask what your thoughts are.. 
Some more info:

I did find that the hashCalc above is pretty much java.lang.string.hashCode().. I have found some weird references to this index creation 

https://forum.search-guard.com/t/tenant-saved-objects-kibana-index/1707 

and 

https://github.com/opendistro-for-elasticsearch/sample-code/issues/182
https://github.com/jcantrill/cluster-logging-tools/blob/master/scripts/kibana-index-name


My Java / TS Skillz are not that good so I'm having a hard time figuring out where the code in the security plugin itself it.. 

There are references I found that do some patterns within index templates for tenants

https://github.com/opendistro-for-elasticsearch/security-kibana-plugin/blob/4c0f50eb2e81ff3c653a0d0f8abe5c78104f9b4e/server/multitenancy/tenant_index.ts#L39-L62

**Comment by nouseforaname** [comment]
Hmm. With that I knowledge I can actually just create a proper datasource pointing to to latest index in use. I guess I can create a request against the endpoint you mentioned for the alias and then just parse out the _XYZ part

**Comment by nouseforaname** [comment]
It seems that the alias only exists once there are multiple versions.. I'm not quite sure yet when exactly a new Version is created. @TonyLovesDevOps  do you have any pointers on what happens before a new index version is created?

**Comment by nouseforaname** [comment]
I created an external datasource implementation for this:

```
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go/aws/session"
	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
)

func main() {

	//external Data source parse Input
	queryBytes, err := ioutil.ReadAll(os.Stdin)
	if err != nil {
		panic(err)
	}
	var query map[string]string
	err = json.Unmarshal(queryBytes, &query)
	if err != nil {
		panic(err)
	}
	tenant := query["tenant"]
	domainUrl := query["domain"]
	region := query["region"]
	aws_profile := query["profile"]

	//Calc Hash
	hashSum := int32(0)
	for _, char := range tenant {
		shift := (hashSum << 5)
		hashSum = (shift - hashSum) + int32(char-0)
	}
	// remove all chars that are not alphanumerical
	regEx, _ := regexp.Compile("[^a-zA-Z0-9]+")
	processedString := regEx.ReplaceAllString(tenant, "")

	index_alias := fmt.Sprintf(".kibana_%v_%v", hashSum, processedString)

	//Use Calculated Index Alias to pull latest index
	sess, err := session.NewSessionWithOptions(session.Options{
		Profile: aws_profile,
	})
	if err != nil {
		panic(err)
	}
	//sign request
	signer := v4.NewSigner(sess.Config.Credentials)
	url := fmt.Sprintf("%v/%v", domainUrl, index_alias)
	req, _ := http.NewRequest(http.MethodHead, url, nil)
	signer.Sign(req, strings.NewReader(""), "es", region, time.Now())

	//fire request
	response, err := http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer response.Body.Close()

	//check if alias exists with head, if it does not yet exist look for index_1 directly
	if response.StatusCode > 299 {
		url = fmt.Sprintf("%v/%v_1", domainUrl, index_alias)
	}

	req, _ = http.NewRequest(http.MethodGet, url, nil)

	signer.Sign(req, strings.NewReader(""), "es", region, time.Now())
	response, err = http.DefaultClient.Do(req)

	//parse data into map
	buf := new(bytes.Buffer)
	buf.ReadFrom(response.Body)
	responseJson := map[string]interface{}{}
	err = json.Unmarshal(
		buf.Bytes(), &responseJson)

	// assumprion is: response object looks like {"index_1": {...}, "index_2": {...}, ...}
	// We only need the keys

	var indices []string
	for index := range responseJson {
		indices = append(indices, index)
	}

	sort.Strings(indices)

	if err != nil {
		panic(err)
	}

	var result = map[string]interface{}{
		"index_alias":    index_alias,
		"index_versions": fmt.Sprintf("%v", indices),
	}

	resultBytes, err := json.Marshal(result)
	if err != nil {
		panic(err)
	}

	os.Stdout.Write(resultBytes)
	os.Exit(0)
}

```

**Comment by nouseforaname** [comment]
was able to create a _2 version of the index via tainting the tenant. 

```      + "index_alias"    = ".kibana_882685070_sectest"
      + "index_versions" = jsonencode(
            [
              + ".kibana_882685070_sectest_2",
            ]
        )
    }
```

**Comment by TonyLovesDevOps** [comment]
> It seems that the alias only exists once there are multiple versions.. I'm not quite sure yet when exactly a new Version is created. @TonyLovesDevOps do you have any pointers on what happens before a new index version is created?
> 
> Can you post an example object you'll see if you query the alias?

TL;DR see point 5 below to see the output of the alias object.

@nouseforaname 🤔 I see an alias created when there is only one tenant index version... here's what I did:

1. Create tenant:
```bash
$ curl -u $ES_USER:$ES_PASSWORD -X PUT -H "Content-Type: application/json" "$ES_URL/_opendistro/_security/api/tenants/foo" -d '{"description":"hack the gibson"}'
{"status":"CREATED","message":"'foo' created."}
```

2. Switch to tenant in Kibana UI (this creates the tenant indices)
3. Show the indices (note only one):
```bash
$ curl -s -u $ES_USER:$ES_PASSWORD -X GET "$ES_URL/_cat/indices/.kibana*foo*?format=json" | jq -r .
[
  {
    "health": "green",
    "status": "open",
    "index": ".kibana_101574_foo_1",
    "uuid": "ABuVCr6kQgaG_w37VGNCrA",
    "pri": "1",
    "rep": "1",
    "docs.count": "1",
    "docs.deleted": "0",
    "store.size": "10kb",
    "pri.store.size": "5kb"
  }
]
```
4. Show the alias:
```bash
$ curl -s -u $ES_USER:$ES_PASSWORD -X GET "$ES_URL/_cat/aliases/.kibana*foo*?format=json" | jq -r .
[
  {
    "alias": ".kibana_101574_foo",
    "index": ".kibana_101574_foo_1",
    "filter": "-",
    "routing.index": "-",
    "routing.search": "-",
    "is_write_index": "-"
  }
]
```
5. Get the alias object (the output is the same as getting the index directly) `curl -s -u $ES_USER:$ES_PASSWORD -X GET "$ES_URL/.kibana_101574_foo" | jq -r .`:
```json
{
  ".kibana_101574_foo_1": {
    "aliases": {
      ".kibana_101574_foo": {}
    },
    "mappings": {
      "dynamic": "strict",
      "_meta": {
        "migrationMappingPropertyHashes": {
          "visualization": "52d7a13ad68a150c4525b292d23e12cc",
          "references": "7997cf5a56cc02bdc9c93361bde732b0",
          "query": "11aaeb7f5f7fa5bb43f25e18ce26e7d9",
          "kql-telemetry": "d12a98a6f19a2d273696597547e064ee",
          "ui-metric": "0d409297dc5ebe1e3a1da691c6ee32e3",
          "application_usage_transactional": "43b8830d5d0df85a6823d290885fc9fd",
          "type": "2f4316de49999235636386fe51dc06c1",
          "url": "c7f66a0df8b1b52f17c28c4adb111105",
          "migrationVersion": "4a1746014a75ade3a714e1db5763276f",
          "sample-data-telemetry": "7d3cfeb915303c9641c59681967ffeb4",
          "index-pattern": "66eccb05066c5a89924f48a9e9736499",
          "search": "5c4b9a6effceb17ae8a0ab22d0c49767",
          "application_usage_totals": "3d1b76c39bfb2cc8296b024d73854724",
          "updated_at": "00da57df13e94e9d98437d13ace4bfe0",
          "search-telemetry": "3d1b76c39bfb2cc8296b024d73854724",
          "namespace": "2f4316de49999235636386fe51dc06c1",
          "timelion-sheet": "9a2a2748877c7a7b582fef201ab1d4cf",
          "config": "c63748b75f39d0c54de12d12c1ccbc20",
          "tsvb-validation-telemetry": "3a37ef6c8700ae6fc97d5c7da00e9215",
          "dashboard": "d00f614b29a80360e1190193fd333bab",
          "namespaces": "2f4316de49999235636386fe51dc06c1"
        }
      },
      "properties": {
        "application_usage_totals": {
          "type": "object",
          "dynamic": "false"
        },
        "application_usage_transactional": {
          "dynamic": "false",
          "properties": {
            "timestamp": {
              "type": "date"
            }
          }
        },
        "config": {
          "dynamic": "false",
          "properties": {
            "buildNum": {
              "type": "keyword"
            }
          }
        },
        "dashboard": {
          "properties": {
            "description": {
              "type": "text"
            },
            "hits": {
              "type": "integer"
            },
            "kibanaSavedObjectMeta": {
              "properties": {
                "searchSourceJSON": {
                  "type": "text"
                }
              }
            },
            "optionsJSON": {
              "type": "text"
            },
            "panelsJSON": {
              "type": "text"
            },
            "refreshInterval": {
              "properties": {
                "display": {
                  "type": "keyword"
                },
                "pause": {
                  "type": "boolean"
                },
                "section": {
                  "type": "integer"
                },
                "value": {
                  "type": "integer"
                }
              }
            },
            "timeFrom": {
              "type": "keyword"
            },
            "timeRestore": {
              "type": "boolean"
            },
            "timeTo": {
              "type": "keyword"
            },
            "title": {
              "type": "text"
            },
            "version": {
              "type": "integer"
            }
          }
        },
        "index-pattern": {
          "properties": {
            "fieldFormatMap": {
              "type": "text"
            },
            "fields": {
              "type": "text"
            },
            "intervalName": {
              "type": "keyword"
            },
            "notExpandable": {
              "type": "boolean"
            },
            "sourceFilters": {
              "type": "text"
            },
            "timeFieldName": {
              "type": "keyword"
            },
            "title": {
              "type": "text"
            },
            "type": {
              "type": "keyword"
            },
            "typeMeta": {
              "type": "keyword"
            }
          }
        },
        "kql-telemetry": {
          "properties": {
            "optInCount": {
              "type": "long"
            },
            "optOutCount": {
              "type": "long"
            }
          }
        },
        "migrationVersion": {
          "dynamic": "true",
          "properties": {
            "config": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            }
          }
        },
        "namespace": {
          "type": "keyword"
        },
        "namespaces": {
          "type": "keyword"
        },
        "query": {
          "properties": {
            "description": {
              "type": "text"
            },
            "filters": {
              "type": "object",
              "enabled": false
            },
            "query": {
              "properties": {
                "language": {
                  "type": "keyword"
                },
                "query": {
                  "type": "keyword",
                  "index": false
                }
              }
            },
            "timefilter": {
              "type": "object",
              "enabled": false
            },
            "title": {
              "type": "text"
            }
          }
        },
        "references": {
          "type": "nested",
          "properties": {
            "id": {
              "type": "keyword"
            },
            "name": {
              "type": "keyword"
            },
            "type": {
              "type": "keyword"
            }
          }
        },
        "sample-data-telemetry": {
          "properties": {
            "installCount": {
              "type": "long"
            },
            "unInstallCount": {
              "type": "long"
            }
          }
        },
        "search": {
          "properties": {
            "columns": {
              "type": "keyword",
              "index": false
            },
            "description": {
              "type": "text"
            },
            "hits": {
              "type": "integer",
              "index": false
            },
            "kibanaSavedObjectMeta": {
              "properties": {
                "searchSourceJSON": {
                  "type": "text",
                  "index": false
                }
              }
            },
            "sort": {
              "type": "keyword",
              "index": false
            },
            "title": {
              "type": "text"
            },
            "version": {
              "type": "integer"
            }
          }
        },
        "search-telemetry": {
          "type": "object",
          "dynamic": "false"
        },
        "timelion-sheet": {
          "properties": {
            "description": {
              "type": "text"
            },
            "hits": {
              "type": "integer"
            },
            "kibanaSavedObjectMeta": {
              "properties": {
                "searchSourceJSON": {
                  "type": "text"
                }
              }
            },
            "timelion_chart_height": {
              "type": "integer"
            },
            "timelion_columns": {
              "type": "integer"
            },
            "timelion_interval": {
              "type": "keyword"
            },
            "timelion_other_interval": {
              "type": "keyword"
            },
            "timelion_rows": {
              "type": "integer"
            },
            "timelion_sheet": {
              "type": "text"
            },
            "title": {
              "type": "text"
            },
            "version": {
              "type": "integer"
            }
          }
        },
        "tsvb-validation-telemetry": {
          "properties": {
            "failedRequests": {
              "type": "long"
            }
          }
        },
        "type": {
          "type": "keyword"
        },
        "ui-metric": {
          "properties": {
            "count": {
              "type": "integer"
            }
          }
        },
        "updated_at": {
          "type": "date"
        },
        "url": {
          "properties": {
            "accessCount": {
              "type": "long"
            },
            "accessDate": {
              "type": "date"
            },
            "createDate": {
              "type": "date"
            },
            "url": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 2048
                }
              }
            }
          }
        },
        "visualization": {
          "properties": {
            "description": {
              "type": "text"
            },
            "kibanaSavedObjectMeta": {
              "properties": {
                "searchSourceJSON": {
                  "type": "text"
                }
              }
            },
            "savedSearchRefName": {
              "type": "keyword"
            },
            "title": {
              "type": "text"
            },
            "uiStateJSON": {
              "type": "text"
            },
            "version": {
              "type": "integer"
            },
            "visState": {
              "type": "text"
            }
          }
        }
      }
    },
    "settings": {
      "index": {
        "number_of_shards": "1",
        "auto_expand_replicas": "0-1",
        "provided_name": ".kibana_101574_foo_1",
        "creation_date": "1620838326729",
        "number_of_replicas": "1",
        "uuid": "ABuVCr6kQgaG_w37VGNCrA",
        "version": {
          "created": "7090199"
        }
      }
    }
  }
}
```

**Comment by nouseforaname** [comment]
@phillbaker I updated my custom data source code above.. It works reliably for me to reference the (latest) index id.. If you want, I'll create a PR with the code embedded somewhere in the tenant resource handling..

**Comment by phillbaker** [comment]
@nouseforaname that'd be great. We could probably store this as a computed attribute on the resource (https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#computed). The "right" way to do it would probably be to build a data source for this, similar to:
```
data "elasticsearch_opendistro_kibana_tenant" "foo" {
  tenant_name   = "test"
}
```

**Comment by phillbaker** [comment]
added in https://github.com/phillbaker/terraform-provider-elasticsearch/commit/df9749c5477f47c48f9ec887aa78469349dc98fe

**Comment by danielmotaleite** [comment]
Sorry for the noise, but i'm failing to understand how to fetch the tenant index name to create a list of index pattern to all tenants

I'm able to create the index pattern if i manually specify the .kibana tenant index... but is there any way to fetch that info from the tenant resource and not manually manage this index names?

```
resource "elasticsearch_kibana_object" "index_pattern" {
  for_each             = var.indexes
  index = ".kibana_-1798208673_test"
  body = <<EOF
[{
    "_id": "index-pattern:${each.key}",
    "_source": {
      "type": "index-pattern",
      "index-pattern": {
        "title": "${reach.key}",
        "timeFieldName": "@timestamp"
      }
    }
}]
EOF
}
```
thanks
